N8N-2827 speed up nodes base

alexgrozav opened the pull request:
## Issue
Building thousands of files in the nodes-base package is slow. Attempt switching to [vite](https://vitejs.dev/) or [esbuild](https://esbuild.github.io/).

## Findings

### Switching to Vite.js
Vite.js is focused on creating application bundles and packages. This means that there should be only a few entry points. The `nodes-base` package has hundreds of entry points that need to be compiled individually. Even though Vite is super fast, it's not what it is built for.

**Vite.js build time**
~~~
npm run build 59.20s user 2.20s system 178% cpu 18.339 total
~~~

### Switching to esbuild
Even after switching to esbuild, I was only able to shave off a few seconds from the build because the `.d.ts` still need to be generated using `tsc`. Esbuild is fast, but it simply ignores typings when compiling https://github.com/evanw/esbuild/issues/95#issuecomment-626069971. 

## Proposed solution
Esbuild is able to compile the whole project in under `1s` without type definitions. The only typings used outside of development are found in the `src` folder. Instead of generating typescript definition files for each node, we can generate them only for the files inside the src folder.

This dramatically reduces build time.

**Old build time**
~~~
npm run build:old  22.30s user 1.10s system 176% cpu 13.229 total
~~~

**New build time**
~~~
npm run build  3.85s user 0.59s system 125% cpu 3.549 total
~~~

> Benchmarks run on a 16" MacBook Pro with M1 Pro

If still needed, we could add a script for generating the `d.ts` for nodes when creating a new release.

Alternatively, could let the `build` command build all `d.ts` files and then have a `build:fast` to rebuild `.ts` using `esbuild`, and copy assets only.

What do you think?

alexgrozav committed:
:sparkles: Migrated nodes-base to esbuild

:construction: Updated dependencies and cleaned up.

:zap: Building nodes-base via esbuild.

:bug: Added copying json files to gulp task.

:recycle: Fixed EditImage esModuleInterop warning and PostMark empty body check warning.

:sparkles: Updated npm-run-all to run in parallel

:recycle: Refactored pg namespace import.

:recycle: Switched to esbuild watch from chokidar.

:recycle: Removed unnecessary params from tsconfig.build.json.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/n8n-io/n8n?pullRequest=3110) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/n8n-io/n8n?pullRequest=3110) before we can accept your contribution.<br/><hr/>**Alex Grozav** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/n8n-io/n8n?pullRequest=3110) it.</sub>

ivov commented on packages/nodes-base/nodes/Postgres/Postgres.node.functions.ts:
```diff
@@ -1,6 +1,6 @@
 import { IDataObject, INodeExecutionData } from 'n8n-workflow';
 import pgPromise from 'pg-promise';
-import pg from 'pg-promise/typescript/pg-subset';
+import * as pg from 'pg-promise/typescript/pg-subset';
```
With `esModuleInterop` enabled, namespace import here should in theory be unnecessary.

alexgrozav replied:
Updated to import `{ IClient }` from pg-subset directly.

ivov commented on packages/nodes-base/package.json:
```diff
@@ -16,13 +16,17 @@
   "scripts": {
     "dev": "npm run watch",
-    "build": "tsc && gulp build:icons && gulp build:translations",
+    "build": "npm-run-all build:esbuild build:assets build:dts",
```
Not very familiar with `npm-run-all`, can we use `--parallel` to build faster?

alexgrozav replied:
Shaved off another second! Thanks for the suggestion! :)

ivov commented on packages/nodes-base/scripts/build.mjs:
```diff
@@ -0,0 +1,38 @@
+		entryPoints: tsFiles,
+		format: 'cjs',
+		outdir: path.resolve(distDir, dir),
+	}).catch(() => process.exit(1));
```
Nitpick: Maybe also log out the build error?

alexgrozav replied:
Done!

ivov commented on packages/nodes-base/scripts/build.mjs:
```diff
@@ -0,0 +1,38 @@
+].forEach(async (dir) => {
+	const tsFiles = await glob(path.resolve(rootDir, dir, '**/*!(.d).ts'));
+
+	esbuild.build({
```
The existing build process outputs source maps. Should esbuild do so as well? Tbh, not sure if we are using them.

alexgrozav replied:
Added sourcemaps generation when building.

ivov commented on packages/nodes-base/scripts/build.mjs:
```diff
@@ -0,0 +1,38 @@
+	'nodes',
+	'src'
+].forEach(async (dir) => {
+	const tsFiles = await glob(path.resolve(rootDir, dir, '**/*!(.d).ts'));
```
Looks like some `.d.ts` files are not getting excluded? `nodes/ActionNetwork/types.d.ts` is included as `dist/nodes/ActionNetwork/types.d.js`

alexgrozav replied:
Updated the build script. Should work fine now!

ivov commented on packages/nodes-base/scripts/watch.mjs:
```diff
@@ -0,0 +1,20 @@
+	shell.exec('npm run build:esbuild');
+}
+
+chokidar.watch([
+	path.resolve(__dirname, '..', 'credentials'),
+	path.resolve(__dirname, '..', 'nodes'),
+	path.resolve(__dirname, '..', 'src'),
+], { ignoreInitial: true })
+	.on('change', listener)
+	.on('add', listener)
+	.on('unlink', listener);
```
1. Curious why a chokidar setup instead of [esbuild watch mode](https://esbuild.github.io/api/#watch)? I expect this rebuilds on changes to any file, not just TS files, i.e. more than what we need?
2. Should we rebuild on `unlink`?
3. If `ignoreInitial` is `false`, it glitches out by re-detecting constantly. If `ignoreInitial` is `true`, then building before watching becomes required, i.e. if we watch without building first, then no `/dist` is created until the first change. This differs from the behavior of `tsc --watch`, which knows to build if there is no `/dist`.
4. Nitpick: Most watch tools give out some feedback e.g. `Listening for file changes...`, but this watch script is silent.

alexgrozav replied:
Good point! Switched to using esbuild in watch mode and added feedback and error handling.

ivov commented on packages/nodes-base/scripts/watch.mjs:
```diff
@@ -0,0 +1,20 @@
+const __filename = fileURLToPath(import.meta.url);
+const __dirname = path.dirname(__filename);
+
+const listener = (event, path) => {
+	shell.exec('npm run build:esbuild');
+}
```
Matter of preference:
```suggestion
const listener = () => {
	shell.exec('npm run build:esbuild');
}
```

ivov commented on packages/nodes-base/tsconfig.build.json:
```diff
@@ -0,0 +1,14 @@
+{
+	"extends": "./tsconfig.json",
+	"compilerOptions": {
+		"emitDeclarationOnly": true,
+		"declaration": true,
+		"outDir": "./dist/src"
+	},
+	"include": [
+		"src/**/*",
+	],
+	"exclude": [
+		"**/*.spec.ts"
+	]
+}
```
1. `declaration: true` is the same as in the parent. Is it necessary here? Same for `exclude` and its contents.
2. `src/**/*` in `include` in the child overrides the parent's. Can we remove it from the parent?
3. The parent is including `nodes/**/*.json` and `credentials/translations/**/*.json` but now `copyAssets` takes care of JSON files. Can we remove these lines from the parent?

alexgrozav replied:
1. Removed `declaration:true` and `exclude`. Thanks for pointing that out!
2. The parent should still be compiling all the files (for generating all `d.ts` files) and targeting all the files (for IDEs, some IDEs such as Webstorm only take the root tsconfig file into consideration).
3. The JSON files still need to be referenced because otherwise you get an error when importing them throughout `nodes-base`.

ivov commented on packages/nodes-base/package.json:
```diff
@@ -16,13 +16,17 @@
+    "build": "npm-run-all build:esbuild build:assets build:dts",
+    "build:esbuild": "node scripts/build.mjs",
+    "build:assets": "gulp build:assets build:translations",
+    "build:dts": "tsc --project tsconfig.build.json",
```
The type `NodeVersionedType` is consumed only once and only in `cli/CredentialsHelper.ts`. Maybe we should relocate this class to the package where it is used and thereby also skip type declaration emitting? (Out of scope, tech debt.)

alexgrozav replied:
Agreed, would be for the best. :)

ivov commented on packages/nodes-base/package.json:
```diff
@@ -707,14 +711,19 @@
     "nodelinter": "^0.1.9",
+    "npm-run-all": "4.1.5",
+    "shelljs": "0.8.5",
+    "tiny-glob": "0.2.9",
```
`tiny-glob` in /nodes-base vs. `fast-glob` in /cli. We should use a single lib - maybe tech debt to switch /cli to `tiny-glob` if that's the one we want to keep.

alexgrozav replied:
I chose `tiny-glob` because it's much faster (~230%) than `fast-glob`. I don't have a strong preference, but the syntax is the same for both. Should be straightforward to switch in `/cli` as well.

ivov reviewed (commented):
Nice work! Cannot wait for 1s build for `/nodes-base`.

ivov reviewed (commented):
Sorry for the delay! Re-review:

1. n8n fails to initialize - unable to load some esbuilt nodes:

```
 git reset --hard; git clean -dffx
lerna bootstrap --hoist; npm run build
npm run start 
```


```
 > n8n@0.168.2 start
> run-script-os

> n8n@0.168.2 start:default
> cd packages/cli/bin && ./n8n

Initializing n8n process
 ›   Error: There was an error: Cannot find module './v1/actions/router'
 ›   Require stack:
 ›   - /Users/ivov/Development/n8n/packages/nodes-base/dist/nodes/BambooHr/BambooHr.node.js
 ›   - /Users/ivov/Development/n8n/packages/cli/dist/src/LoadNodesAndCredentials.js
 ›   - /Users/ivov/Development/n8n/packages/cli/dist/src/index.js
 ›   - /Users/ivov/Development/n8n/packages/cli/dist/commands/execute.js
 ›   - /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/plugin.js
 ›   - /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/config.js
 ›   - /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/index.js
 ›   - /Users/ivov/Development/n8n/node_modules/@oclif/command/lib/command.js
 ›   - /Users/ivov/Development/n8n/node_modules/@oclif/command/lib/index.js
 ›   - /Users/ivov/Development/n8n/packages/cli/bin/n8n
```

Removing Bamboo HR triggers same error for Cisco Webex and then Elasticsearch.

2. `.d.ts` are still being included, becoming `.d.js` in /dist. The glob in `build.mjs` is not excluding them: `console.log(tsFiles.filter(f => f.endsWith('.d.ts')));`
3. The existing TS build process emits CJS with the use strict directive; esbuild is [not doing this](https://github.com/evanw/esbuild/issues/422), but supports a [workaround](https://esbuild.github.io/api/#banner). 
4. Switching to a non-English locale causes the build to fail:

```
export N8N_DEFAULT_LOCALE=de
cd packages/nodes-base; npm run build
```

```
node:events:368
      throw er; // Unhandled 'error' event
      ^

Error: ENOENT: no such file or directory, open '/Users/ivov/Development/n8n/packages/nodes-base/dist/nodes/headers.js'
Emitted 'error' event on Domain instance at:
    at emit (node:internal/process/promises:134:35)
    at processPromiseRejections (node:internal/process/promises:242:25)
    at processTicksAndRejections (node:internal/process/task_queues:97:32) {
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: '/Users/ivov/Development/n8n/packages/nodes-base/dist/nodes/headers.js'
}
ERROR: "build:assets" exited with 1.
```

This issue is absent from `master`. The build stages are now parallel, so I expect the attempt to write to the headers path finishes first, failing to find the dist dir.

Edit: 4.5. Should we restore type-checking to the build process with [`--noEmit`](https://www.typescriptlang.org/tsconfig/#noEmit)?

## Minor details

5. `'\n[esbuild] Watching for changes...'` is shown both for watch and for build, instead of just for watch. You can use the `watch.onRebuild` arg, refer to the second watch mode example:

```js
require('esbuild').build({
  entryPoints: ['app.js'],
  outfile: 'out.js',
  bundle: true,
  watch: {
    onRebuild(error, result) {
      if (error) console.error('watch build failed:', error)
      else console.log('watch build succeeded:', result)
    },
  },
}).then(result => {
  console.log('watching...')
})
```

6. Instead of removing the existing watch script reference in `package.json`, we should keep it as `watch:old` just like the old build script. In a few weeks if there are no issues, we can remove both.
7. After deletion of `watch.mjs`, `shelljs` is now only used for `rm -rf`. We already have `rimraf`, so `shelljs` might not be needed.
8. The only JSON files in /nodes-base/nodes and /nodes-base/credentials are codex files `.node.json` and potentially translation files. Both kinds are required cross-package from /cli, i.e. they are not directly imported inside /nodes-base, so I expect that `nodes/**/*.json` and `credentials/translations/**/*.json` should be removable from `tsconfig.json` without import issues.

```
find ./packages/nodes-base -type f -name \*.json ! -name \*.node.json
```

9. (Nitpick) Naming of `tsconfig.build.json` - the child only takes care of building /src, but both parent and child are responsible for the entire build, which makes the `.build.` infix for /src potentially confusing. This will become moot once we move src into /cli and skip declaration generation altogether.
10. (Out of scope, tech debt.) We should eventually explore the esbuild minify and tree-shaking flags.

alexgrozav committed:
:sparkles: Updated esbuild configuration for nodes-base.

alexgrozav commented:
> Sorry for the delay! Re-review:
> 
> 1. n8n fails to initialize - unable to load some esbuilt nodes:
> 
> ```
>  git reset --hard; git clean -dffx
> lerna bootstrap --hoist; npm run build
> npm run start 
> ```
> 
> ```
>  > n8n@0.168.2 start
> > run-script-os
> 
> > n8n@0.168.2 start:default
> > cd packages/cli/bin && ./n8n
> 
> Initializing n8n process
>  ›   Error: There was an error: Cannot find module './v1/actions/router'
>  ›   Require stack:
>  ›   - /Users/ivov/Development/n8n/packages/nodes-base/dist/nodes/BambooHr/BambooHr.node.js
>  ›   - /Users/ivov/Development/n8n/packages/cli/dist/src/LoadNodesAndCredentials.js
>  ›   - /Users/ivov/Development/n8n/packages/cli/dist/src/index.js
>  ›   - /Users/ivov/Development/n8n/packages/cli/dist/commands/execute.js
>  ›   - /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/plugin.js
>  ›   - /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/config.js
>  ›   - /Users/ivov/Development/n8n/node_modules/@oclif/config/lib/index.js
>  ›   - /Users/ivov/Development/n8n/node_modules/@oclif/command/lib/command.js
>  ›   - /Users/ivov/Development/n8n/node_modules/@oclif/command/lib/index.js
>  ›   - /Users/ivov/Development/n8n/packages/cli/bin/n8n
> ```
> 
> Removing Bamboo HR triggers same error for Cisco Webex and then Elasticsearch.

The issue seems to have been `tiny-glob` not going deeply enough into the directory tree. Switched to `fast-glob` to match `cli`.

> 2. `.d.ts` are still being included, becoming `.d.js` in /dist. The glob in `build.mjs` is not excluding them: `console.log(tsFiles.filter(f => f.endsWith('.d.ts')));`

Fixed. Fast glob supports negation glob.

> 3. The existing TS build process emits CJS with the use strict directive; esbuild is [not doing this](https://github.com/evanw/esbuild/issues/422), but supports a [workaround](https://esbuild.github.io/api/#banner).

Added `'use strict';` via `banner.js`.

> 4. Switching to a non-English locale causes the build to fail:
> 
> ```
> export N8N_DEFAULT_LOCALE=de
> cd packages/nodes-base; npm run build
> ```
> 
> ```
> node:events:368
>       throw er; // Unhandled 'error' event
>       ^
> 
> Error: ENOENT: no such file or directory, open '/Users/ivov/Development/n8n/packages/nodes-base/dist/nodes/headers.js'
> Emitted 'error' event on Domain instance at:
>     at emit (node:internal/process/promises:134:35)
>     at processPromiseRejections (node:internal/process/promises:242:25)
>     at processTicksAndRejections (node:internal/process/task_queues:97:32) {
>   errno: -2,
>   code: 'ENOENT',
>   syscall: 'open',
>   path: '/Users/ivov/Development/n8n/packages/nodes-base/dist/nodes/headers.js'
> }
> ERROR: "build:assets" exited with 1.
> ```
> 
> This issue is absent from `master`. The build stages are now parallel, so I expect the attempt to write to the headers path finishes first, failing to find the dist dir.
> 
> Edit: 4.5. Should we restore type-checking to the build process with [`--noEmit`](https://www.typescriptlang.org/tsconfig/#noEmit)?
> 

Fixed. Ensured folder creation upon build. Extracted cleanup phase to start before parallel execution (duh! 😰).

> ## Minor details
> 5. `'\n[esbuild] Watching for changes...'` is shown both for watch and for build, instead of just for watch. You can use the `watch.onRebuild` arg, refer to the second watch mode example:
> 
> ```js
> require('esbuild').build({
>   entryPoints: ['app.js'],
>   outfile: 'out.js',
>   bundle: true,
>   watch: {
>     onRebuild(error, result) {
>       if (error) console.error('watch build failed:', error)
>       else console.log('watch build succeeded:', result)
>     },
>   },
> }).then(result => {
>   console.log('watching...')
> })
> ```
> 

Added proper build message for build mode and watch mode. 

> 6. Instead of removing the existing watch script reference in `package.json`, we should keep it as `watch:old` just like the old build script. In a few weeks if there are no issues, we can remove both.



> 7. After deletion of `watch.mjs`, `shelljs` is now only used for `rm -rf`. We already have `rimraf`, so `shelljs` might not be needed.

Removed `shelljs`.

> 8. The only JSON files in /nodes-base/nodes and /nodes-base/credentials are codex files `.node.json` and potentially translation files. Both kinds are required cross-package from /cli, i.e. they are not directly imported inside /nodes-base, so I expect that `nodes/**/*.json` and `credentials/translations/**/*.json` should be removable from `tsconfig.json` without import issues.
> 
> ```
> find ./packages/nodes-base -type f -name \*.json ! -name \*.node.json
> ```
> 

Removed irrelevant files from tsconfig.json.

> 9. (Nitpick) Naming of `tsconfig.build.json` - the child only takes care of building /src, but both parent and child are responsible for the entire build, which makes the `.build.` infix for /src potentially confusing. This will become moot once we move src into /cli and skip declaration generation altogether.

Renamed it to `tsconfig.dts.json`. Does this make more sense? How would you name it?

> 10. (Out of scope, tech debt.) We should eventually explore the esbuild minify and tree-shaking flags.

Definitely! I'll create a task for it after this one gets merged.

alexgrozav committed:
:recycle: Updated nodes-base watch mode messages.

ivov commented:
Thanks for addressing my comments :+1:

1. `.d.ts` are still being included, becoming `.d.js` in `/dist`. Example: `packages/nodes-base/nodes/ActionNetwork/types.d.js` A sourcemap is also being generated for the transpiled type def file: `types.d.js.map`
2. The esbuild dist size is 32.2 MB, the tsc dist was 23.9 MB. Helpers like `__toCommonJS` are repeated in every file in every node. Is this size increase accounted for and tolerable? More info: https://github.com/evanw/esbuild/issues/1984 
3. Typechecking is no longer part of the build process. Is this intentional?
4. Re: naming for `tsconfig.dts.json`, I was thinking of `tsconfig.src.json`, since `/src` is the only dir that it targets.

alexgrozav committed:
fix: Fixed package-lock merge conflicts.

alexgrozav committed:
feat: Added typecheck to build command. Added build:fast command.

alexgrozav committed:
fix: Fixed type issues and sped up typecheck.

alexgrozav commented:
> Thanks for addressing my comments 👍
> 
> 1. `.d.ts` are still being included, becoming `.d.js` in `/dist`. Example: `packages/nodes-base/nodes/ActionNetwork/types.d.js` A sourcemap is also being generated for the transpiled type def file: `types.d.js.map`

Looks like the `fast-glob` block pattern is not working properly. Filtered using JS.

> 2. The esbuild dist size is 32.2 MB, the tsc dist was 23.9 MB. Helpers like `__toCommonJS` are repeated in every file in every node. Is this size increase accounted for and tolerable? More info: [QUESTION: best way use esbuild to bundle TypeScript? evanw/esbuild#1984](https://github.com/evanw/esbuild/issues/1984)

Not much I can do about this, unfortunately. 

> 3. Typechecking is no longer part of the build process. Is this intentional?

- Added typechecking back when running `npm run build`.

> 4. Re: naming for `tsconfig.dts.json`, I was thinking of `tsconfig.src.json`, since `/src` is the only dir that it targets.


----

- Added `npm run build:fast` which skips type checking and dramatically reduces build time
- I found a way to speed `tsc` up. 
  - Apparently, if you don't refer a project when running it from the command line, you're skipping a lot of boilerplate code. This resulted in a `2s` increase.
  - More than that, if you add `--skipLibCheck`, you shave off another `3s` from the build time at the cost of ignoring `d.ts` file dependencies with errors.

---

~~~
npm run build  16.37s user 1.51s system 235% cpu 7.577 total
~~~

~~~
npm run build:fast  1.35s user 0.82s system 139% cpu 1.561 total
~~~

~~~
npm run build:old  21.49s user 0.99s system 180% cpu 11.919 total
~~~

ivov reviewed (commented):
Thank you for the changes. Some final comments:

1. It is unusual that we only need to build `/nodes-base` in isolation. `/nodes-base` is almost always built together with all other packages  via `dev` in the root `package.json`, i.e. `lerna exec npm run dev --parallel`, which will run `npm run build` in `/nodes-base`. This means that `npm run build:fast` is unlikely to be used in day-to-day work. Should we add a root command for fast builds?

2. `skipLibCheck` will also skip checking of declaration files in the project itself, which is relevant because many nodes contain custom `.d.ts` files. The best of both worlds would be enabling this flag for the speed boost, but also converting all `.d.ts` files to `.ts` to ensure they're type-checked. Context: [1](https://github.com/microsoft/TypeScript/wiki/Performance#skipping-dts-checking) [2](https://www.typescriptlang.org/tsconfig#skipLibCheck) [3](https://www.testim.io/blog/typescript-skiplibcheck/) [4](https://twitter.com/orta/status/1260583041740812288) Maybe we should consider delaying this flag until we can take the time to convert them.

3. `build.mjs` is throwing off ESLint:

```
Parsing error: "parserOptions.project" has been set for @typescript-eslint/parser.
The file does not match your project config: packages/nodes-base/scripts/build.mjs.
The extension for the file (.mjs) is non-standard. You should add "parserOptions.extraFileExtensions" to your config.
```

From a quick check, mjs needs to be [added to `extraFileExtensions`](https://github.com/typescript-eslint/typescript-eslint/issues/3383) or converted to `mts` which is supported by default, or else we could disregard mjs in `.eslintignore`.

4. `build:typecheck` is now inlining multiple flags to avoid using `--project`. These flags are a mix of settings already present in `tsconfig.json` e.g. `module`  and of settings absent from `tsconfig.json`, e.g. `downlevelIteration` and `allowSyntheticDefaultImports`. I am reluctant to introduce duplication and add new compiler flags that were not needed before and that required changes to the source as in Magento and Action Network, for a 2s gain. Commenting mostly to bring up the concern, but up to you.

5. Why switch `module` and `target` to ESNext? From the TS docs: `The special ESNext value refers to the highest version your version of TypeScript supports. This setting should be used with caution, since it doesn’t mean the same thing between different TypeScript versions and can make upgrades less predictable.`

6. `fast-glob` is listed twice in `nodes-base/package.json`, as regular dependency and as dev dependency.

7. `"test/**/*"` is being included and `"**/*.spec.ts"` is being excluded but all tests in `/nodes-base` are JS.

alexgrozav committed:
feat: Parallelized build commands. Added build:fast.

alexgrozav committed:
refactor: Moved NodeVersionedType to packages/workflow to allow parallelization.

alexgrozav committed:
feat: Migrated from lerna to turborepo.

alexgrozav committed:
chore: Updated package-lock.json

alexgrozav committed:
fix: Fixed target and module in nodes-base tsconfig. Removed dependency double-listing.

alexgrozav committed:
fix: Fixed linting issues.

alexgrozav committed:
chore: Merged master into N8N-2827-speed-up-nodes-base.

alexgrozav committed:
chore: Updated new nodes-base package.json dependencies to use ~.

alexgrozav commented on packages/nodes-base/package.json:
```diff
@@ -713,11 +717,14 @@
     "n8n-workflow": "~0.101.0",
     "nodelinter": "^0.1.9",
+    "npm-run-all": "4.1.5",
+    "rimraf": "3.0.2",
```
Thanks for suggesting. Done!

alexgrozav commented on package.json:
```diff
@@ -22,8 +22,13 @@
   },
   "devDependencies": {
     "lerna": "^3.13.1",
```
Unfortunately, we still need the bootstrapping feature from lerna until we move officially to npm workspaces.

ivov reviewed (commented):
Nice idea to bring in Turborepo :)

1. Can you please run these? Before making any changes from the review.

```
git reset --hard; git clean -dffx
lerna bootstrap --hoist
time npm run build
// make a change in a file
time npm run build
system_profiler SPHardwareDataType
```

I get 2m 10s (first build) and 33s (cached build) on:

```
Model Name: MacBook Pro
Model Identifier: MacBookPro17,1
Chip: Apple M1
Total Number of Cores: 8 (4 performance and 4 efficiency)
Memory: 16 GB
```

Comparing against my [baseline benchmarks](https://linear.app/n8n/issue/N8N-3516#comment-104eaa6d) without esbuild and without turbo, only the cached build shows gains, but the first build is identical - this is surprising, as switching to esbuild in `nodes-base` should give us some gain.

2. We are removing `"nodes/**/*.json"` from `packages/nodes-base/tsconfig.json`, so with this new setup we no longer have codex files in dist. Therefore we should confirm that no external services rely on codex files in dist. For example, [`Supported Actions`](https://n8n.io/integrations/494-action-network/) and [`Basic Operations`](https://docs.n8n.io/integrations/nodes/n8n-nodes-base.actionnetwork/) are likely reading from codex files. Bringing this up mostly as I do not know offhand whether they would be affected - if they are reading from "untranspiled" codex files, then no issue.

alexgrozav committed:
fix: Updated turborepo config.

alexgrozav committed:
chore: Updated imports. Removed nodes-base/src references.

alexgrozav commented on package.json:
```diff
@@ -22,8 +22,13 @@
   },
   "devDependencies": {
     "lerna": "^3.13.1",
+    "npm-run-all": "4.1.5",
```
Good catch! Removed.

alexgrozav commented on packages/nodes-base/tsconfig.json:
```diff
@@ -12,23 +12,18 @@
 		"outDir": "./dist/",
 		"target": "es2019",
-		"sourceMap": true,
+		"sourceMap": false,
```
Thank you. Removed.

alexgrozav commented on packages/nodes-base/package.json:
```diff
@@ -16,15 +16,19 @@
+    "build": "rimraf dist && npm-run-all --parallel build:esbuild build:gulp build:tsc",
+    "build:esbuild": "node scripts/build.mjs",
+    "build:gulp": "gulp build:assets build:translations",
+    "build:tsc": "tsc --project tsconfig.json --emitDeclarationOnly",
```
1. No need to specify project.
2. Added d.ts emitting since it doesn't impact build speed.

alexgrozav commented on packages/nodes-base/gulpfile.js:
```diff
@@ -1,16 +1,18 @@
 const PURPLE_ANSI_COLOR_CODE = 35;
 
-task('build:icons', copyIcons);
+task('build:assets', copyAssets);
```
Already done in the `package.json` script. `gulp build:assets build:translations` will run the two tasks in parallel.

alexgrozav commented on packages/nodes-base/scripts/build.mjs:
```diff
@@ -0,0 +1,45 @@
+ */
+
+const tsFiles = (await glob([path.resolve(rootDir, '{credentials,nodes,src}/**/*.ts')]))
+	.filter((file) => !file.endsWith('.d.ts'));
```
1. Removed `src`
2. That's essentially what the glob does as well, and it runs only once. The overhead is `tsc`, unfortunately. The `build.mjs` script runs in a few hundred milliseconds.

ivov replied:
2. But after glob there is filter so double iteration? `walk` is intended to avoid double iteration as it includes `filter` as `test`. Let me know if I'm misunderstanding.

alexgrozav commented on packages/workflow/src/Interfaces.ts:
```diff
@@ -7,11 +7,11 @@ import * as express from 'express';
-import { WorkflowActivationError } from './WorkflowActivationError';
-import { WorkflowOperationError } from './WorkflowErrors';
-import { NodeApiError, NodeOperationError } from './NodeErrors';
+import type { Workflow } from './Workflow';
+import type { WorkflowHooks } from './WorkflowHooks';
+import type { WorkflowActivationError } from './WorkflowActivationError';
+import type { WorkflowOperationError } from './WorkflowErrors';
+import type { NodeApiError, NodeOperationError } from './NodeErrors';
```
Updated to `import type`.

alexgrozav commented on packages/workflow/src/NodeVersionedType.ts:
```diff
@@ -1,25 +1,29 @@
-import { INodeType, INodeTypeBaseDescription, INodeVersionedType } from 'n8n-workflow';
+import { INodeType, INodeTypeBaseDescription, INodeVersionedType } from './Interfaces';
```
Updated.

alexgrozav commented on packages/workflow/src/NodeVersionedType.ts:
```diff
@@ -1,25 +1,29 @@
 	description: INodeTypeBaseDescription;
 
-	constructor(nodeVersions: INodeVersionedType['nodeVersions'], description: INodeTypeBaseDescription) {
+	constructor(
```
I've had the same concern. But after giving it some thought, the `NodeVersionedType.ts` file doesn't look like something that's likely to change too often, and it directly implements an interface declared in `workflow`, so the location felt right. 

In my opinion the build speed benefit will be worth it in the long run. What do you think?

alexgrozav commented on turbo.json:
```diff
@@ -0,0 +1,41 @@
+{
+	"$schema": "https://turborepo.org/schema.json",
```
Looks like there's no real benefit in defining build dependencies explicitly. When left to use the dependency graph, `turborepo` does a great job by itself.

1. You're right. It's able to infer the build outputs.
2. I have not looked into `test` `format`, `lint` and `lintfix` yet. Will do now.
3. Good point. Removed.
4. Removed `outputMode`.
5. Removed `inputs`
6. Increased memory limit, but doesn't impact build speed unfortunately. Webpack / Vue CLI is still a bottleneck until we migrate to Vite.js.

alexgrozav replied:
Converted remaining `lerna run` scripts.

alexgrozav committed:
feat: Added turborepo format and lint commands.

alexgrozav committed:
feat: Converted lerna run commands to turborepo.

alexgrozav committed:
fix: Reverted fork-ts-checker plugin changes.

alexgrozav committed:
chore: Merged master into N8N-2827-speed-up-nodes-base.

alexgrozav committed:
fix: Bumped package.json version to invalidate ci cache.

alexgrozav committed:
fix: Reverted version change.

fix: Added turbo-linux-64 to CI step.

x

alexgrozav committed:
fix: Fixed turborepo version.

alexgrozav committed:
fix: Disabled npm caches in GitHub actions.

alexgrozav committed:
fix: Updated package-lock.json

alexgrozav committed:
chore: Merged master into N8N-2827-speed-up-nodes-base.

alexgrozav committed:
Merge branch 'master' into N8N-2827-speed-up-nodes-base

alexgrozav committed:
Merge branch 'master' into N8N-2827-speed-up-nodes-base

alexgrozav committed:
Merge branch 'master' into N8N-2827-speed-up-nodes-base

alexgrozav committed:
chore: Merged master into N8N-2827-speed-up-nodes-base

alexgrozav committed:
chore: Reverted cache: false from CI.
